scripts/train_lm.py

```diff
@@ -145,14 +145,65 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
     f.close()
 
 def FormatMetaparameters(metaparameters):
+    # store metaparameters as floats during checking and rerounding
     out = []
+    # final output metaparameters in string
+    out_param = []
+    # round all float metaparameters to 3rd decimal places 
+    round_index = 3
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
+        x = round(param, round_index)
         out.append(x)
-
-    return ','.join(out)
+    
+    size = len(metaparameters)
+    
+    # check1: whether 0.000 or 1.000 occur because of rounding
+    # after this step, no 0.000 or 1.000 occurs unless the original number equal
+    # to 0 or 1 (if this happen, the parameters are invalid and it exits 
+    # with a warning, while this case can be very rare)
+    for i in range(0, size): 
+        if out[i] in [0.0, 1.0]:
+            out[i] = RoundToProperDigit(metaparameters[i], round_index)
+            if out[i] in [0.0, 1.0]:
+                sys.exit("train_lm.py: there is a {0}. metaparameters should be in range (0, 1).".format(out[i]))
+    
+    # check2: check repeating values. if happen, then reround them until they are different
+    # if there exist same values originally, exit with a warning (this can be very rare) 
+    # after this step, there should be no any repeating values in set out
+    for i in range(0, size):
+        x = out[i]
+        for j in range(i + 1, size):
+            y = out[j]
+            # we will round to the next digit when we detect repeated values
+            round_digit = round_index + 1
+            round_max_digit = max(len(str(metaparameters[i])) - 2, len(str(metaparameters[j])) -2)
+            while x == y and round_digit <= round_max_digit:
+                x = round(metaparameters[i], round_digit)
+                y = round(metaparameters[j], round_digit)
+                round_digit += 1
+            # actually validate_metaparameters.py will validate those parameters later. 
+            # except for satisfying condition that those parameters should be in 
+            # range (0, 1), parameters D1, D2, D3, D4 for a certain xgram model 
+            # should satisfy D1 > D2 > D3 > D4. 
+            # here we only checked the repeating values but not the order relationship
+            if x == y:
+                sys.exit("train_lm.py: {0} and {1} are same. metaparameters can not be exactly same.".format(x,y))
```

Comment:
I believe this is not an error unless x and y are both metaparameters representing data-type weights.
So this could cause failure when there should be no failure.